### DISCUSSION TITLE
AG Grid Enterprise

### DISCUSSION THREAD

**Discussion by ducnva** [zauberzeug/nicegui#2431] [Ideas / Feature Requests]
### Question

Hi everyone,
How can I use AG Grid Enterprise?.
How Can I use/setup key for ag-grid enterprise

**Comment by falkoschindler** [comment]
Currently NiceGUI only supports AG Grid Community and I'm not sure how AG Grid Enterprise could be integrated.
Let's convert this discussion into a feature request. Maybe someone from the community has an idea how to add this feature. 🙂

**Comment by luongnv275** [comment]
I am a user of NiceGUI and have been enjoying its capabilities with AG Grid Community. However, I believe that integrating AG Grid Enterprise would bring additional value to the tool.

AG Grid Enterprise offers advanced features that could greatly enhance the functionality of NiceGUI. Some specific features I am interested in include [mention features], and I believe many other users share this interest.

Considering the growing popularity of AG Grid Enterprise, it would be fantastic if NiceGUI could provide support for it

**Comment by xaptronic** [comment]
Here is a pull request that adds the ability to use aggrid in enterprise mode: https://github.com/zauberzeug/nicegui/pull/3757

Would it be helpful to additionally create a feature request issue for this?

**Reply by rodja** [reply]
The PR is sufficient. No need for another feature request. Thanks for looking into this. We will review it as soon as poosible.

**Comment by xaptronic** [comment]
Here is my initial aggrid enterprise component:

https://github.com/xaptronic/nicegui-aggrid-enterprise

Likely will make some additional updates, but if anything obvious appears missing feel free to mention.

**Reply by falkoschindler** [reply]
Nice!

**Reply by jtrungle** [reply]
I've found this useful so thank you! Wondering if there's any chance of including Ag-Charts being imported too to get Sparklines in aggrid https://www.ag-grid.com/javascript-data-grid/sparklines-installation/

**Reply by xaptronic** [reply]
> I've found this useful so thank you! Wondering if there's any chance of including Ag-Charts being imported too to get Sparklines in aggrid https://www.ag-grid.com/javascript-data-grid/sparklines-installation/

Hey @falkoschindler would you have any insight on whether the JS loaded by a component could be dynamically loaded based on a parameter?

Right now it uses "dependencies" array, but perhaps it could use loadResource instead to optionally load the charts version instead?

Alternatively, there could be a subclass of the enterprise aggrid that overrides the js that's loaded..

Curious if you have any opinion about which might be a better approach?

**Reply by falkoschindler** [reply]
Hi @xaptronic,

Adding a subclassing parameter to switch between different dependencies seems like a viable solution:

```py
class MyElement(ui.element):

    def __init_subclass__(cls, param: str) -> None:
        super().__init_subclass__(dependencies=['lib/a.js'] if param == 'A' else ['lib/b.js'])

class A(MyElement, param='A'):
    pass

class B(MyElement, param='B'):
    pass
```

But registering resources during instantiation using `add_resource` is also a good option, I guess. See https://github.com/zauberzeug/nicegui/blob/main/nicegui/elements/leaflet.py#L44 and https://github.com/zauberzeug/nicegui/blob/main/nicegui/elements/codemirror.py#L284 for example.

Which one to choose depends on whether a subclass should always have certain dependencies, or if individual instances should be able to have different dependencies.

**Comment by amks1** [comment]
@xaptronic That's an awesome library, thanks.
Not related to enterprise but do you know how I can create custom components in ag-grid for use in NiceGUI? Specifically I'm trying to make a custom cellEditor. 
I tried doing `ui.run_javascript` with the component code inside it but I may be doing it wrong and my lack of JS knowledge is a bottleneck.

**Reply by falkoschindler** [reply]
Thanks, @xaptronic!
Regarding custom AG Grid components, please create a new Q&A discussion and share an example of what you tried and what did not work.

**Comment by gioxc88** [comment]
Can we just merge this extensions into nicegui ?

Ipyaggrid (which I don't think it's actively maintained anymore) also has the option to use enterprise features.

For testing you don't even need an enterprise licence btw

**Reply by python-and-novella** [reply]
Check this:https://github.com/zauberzeug/nicegui/pull/3757#issuecomment-2393321283

**Comment by falkoschindler** [comment]
**Update: Renewed interest in AG Grid Enterprise support**

This topic has come up again in #5623, so I wanted to provide a status update and outline possible paths forward.

**Current situation:**

- NiceGUI ships with AG Grid Community (currently v34.2.0), which covers many use cases but lacks Enterprise features like row grouping, Excel export, pivoting, and server-side row model.
- PR #3757 attempted to add Enterprise support but was closed due to licensing constraints and maintenance concerns.
- The community extension [xaptronic/nicegui-aggrid-enterprise](https://github.com/xaptronic/nicegui-aggrid-enterprise) showed the technical approach works, but appears to be unmaintained.

**Why we can't include Enterprise in core:**

AG Grid Enterprise is commercial software requiring paid licenses. We can't bundle it in NiceGUI's distribution - users must bring their own license.

**Possible paths forward:**

1. **Revive the community extension**: If @xaptronic is interested in continuing maintenance, or if someone else wants to take over, we'd be happy to link to it from our documentation and help promote it.

2. **Better documentation/tooling**: We could provide guidance or examples showing how users with an Enterprise license can integrate it into their NiceGUI apps.

**Call to action:**

- If you're using AG Grid Enterprise with NiceGUI today, how are you doing it? Share your approach!
- Which Enterprise features are most critical for your use cases?
- Anyone interested in maintaining a community extension?

Your feedback helps us understand the demand and prioritize our efforts. Thanks!

**Comment by xaptronic** [comment]
Two comments:
1. I'm doing an upgrade of that aggrid-enterprise package to bring it up to date with where NiceGUI is at these days.
2. The enterprise package is largely a copy of the community one that is built in, but with enterprise packages instead. To that end, option 2) of docs / tooling for using the enterprise packages with the built in package might make more sense, given the aggrid packages are basically the same, just support different feature sets.

**Comment by evnchn** [comment]
I don't have an AG Grid enterprise license, and I think that fact is **the** bottleneck. 

How can I test something which I cannot run?

**Reply by evnchn** [reply]
And on that: @xaptronic how did you do the devwork with that extension? Do you have a license or do you have a workaround?

**Comment by gioxc88** [comment]
You shouldn't need an AGGrid enterprise license to test it.

AG Grid Enterprise licensing works by requiring a paid key for production use, but you can test its advanced features (like grouping, pivoting, master/detail) locally without one; they display a console warning and a temporary watermark to indicate unlicensed use, encouraging you to get a trial or purchase. To test, just install the ag-grid-enterprise package and use the features, which will work until you deploy to production, where the watermarks/errors appear unless a valid license is provided.

I think it's safe to add the package to nicegui

It should NOT be a niceGUI problem if people use it for production without a valid licence, as niceGUI is just providing the interface.

Ipyaggrid for example does exactly this.
It provides all the enterprise features out of the box; with the option to pass a licence key at initialisation

You can literally install the enterprise from npm and start using it right away

I am not sure why you are against the idea to add it to nicegui core

**Reply by falkoschindler** [reply]
Thanks for the suggestion @gioxc88, but unfortunately this isn't an option for us.

I looked into the [AG Grid Enterprise EULA](https://www.ag-grid.com/eula/AG-Grid-Enterprise-License-Latest.html), and **clause 3.7(b)** explicitly prohibits redistribution as part of:

> *"any application that can be described as a development toolkit or library, an application builder, a website builder, a user interface designer, or any application intended for use by software, application, or website developers or designers"*

NiceGUI is clearly a development toolkit/library, so bundling AG Grid Enterprise would directly violate their license terms - regardless of whether end users have their own licenses.

**Reply by gioxc88** [reply]
I see, I was too naive .

In this case I don't see how an extension or separate library would solve this problem 
Wouldn't they violate the licence ?

**Reply by evnchn** [reply]
They just don't allow redistribution. 

Can I make it into bring-your-own-aggrid?

**Reply by falkoschindler** [reply]
> Wouldn't they violate the licence?

Good point! If clause 3.7(b) prohibits redistribution as part of a "development toolkit or library," then any extension that bundles AG Grid Enterprise would face the same issue.

I could imagine, though, that the licence doesn't have non-JS platforms like Jupyter and NiceGUI in mind and that a standalone package like nicegui-aggrid-enterprise could risk being in a kind of grey area. A large framework like NiceGUI, however, should remain extra clean and cautious in this regard to avoid being banned in corporate environments for legal reasons.

The license-compliant approach for NiceGUI would be to provide glue code and documentation only, and have users install ag-grid-enterprise themselves directly from AG Grid with their own license.

> Can I make it into bring-your-own-aggrid?

Not quite sure what you mean...

**Comment by evnchn** [comment]
#5629

**Comment by xaptronic** [comment]
If interested:

- Updates current nicegui-aggrid-enterprise to latest version and nicegui 3.0 API - https://github.com/xaptronic/nicegui-aggrid-enterprise/pull/6
- Updates nicegui-aggrid-enterprise to have a build command that builds does npm install on the users machine instead of bundling it: https://github.com/xaptronic/nicegui-aggrid-enterprise/pull/5

**Comment by falkoschindler** [comment]
In version 3.6 we just released PR #5629 which simplifies the use of AG Grid Enterprise:

```py
ui.aggrid.set_esm_module('https://cdn.jsdelivr.net/npm/ag-grid-enterprise@35.0.0/+esm')

...

ui.aggrid(..., modules='enterprise')
```

This is probably as far as we can push it legally, so I'm closing this feature request.
